cmd/juju/application/deployer/deployer.go

```diff
@@ -54,25 +75,202 @@ func NewDeployerFactory(dep DeployerDependencies) DeployerFactory {
+				return nil, errors.Trace(err)
+			} else if errors.Cause(err) == zip.ErrFormat {
+				return nil, errors.Errorf("invalid charm or bundle provided at %q", charmOrBundle)
+			} else if errors.Is(err, errors.NotFound) {
+				return nil, errors.Wrap(err, errors.NotFoundf("charm or bundle at %q", charmOrBundle))
+			} else if err != nil && err != os.ErrNotExist {
+				// If we get a "not exists" error then we attempt to interpret
+				// the supplied charm reference as a URL elsewhere, otherwise
+				// we return the error.
+				return nil, errors.Annotatef(err, "attempting to deploy %q", charmOrBundle)
+			} else if err != nil {
+				// Pre-deployed local charm?
+				logger.Debugf("cannot interpret as local charm: %v, attempting to deploy pre-deployed charm", err)
+				// If the charm's schema is local, we should definitively attempt
+				// to deploy a charm that's already deployed in the
+				// environment.
+				if userCharmURL.Schema != "local" {
+					return nil, errors.Errorf("cannot interpret as a redeployment of a local charm from the controller")
+				}
+				return &predeployedLocalCharm{
+					deployCharm:  d.newDeployCharm(),
+					userCharmURL: userCharmURL,
+				}, nil
+			} else {
+				// If all above goes well, set the deployer kind to localCharmDeployerType
+				dk = &localCharmDeployerType{seriesName, imageStream, ch, curl}
+			}
+		}
+	} else {
+		// Repository bundle?
+		// To deploy by revision, the revision number must be in the origin for a charmhub bundle
+		if userCharmURL.Revision != -1 && charm.CharmHub.Matches(userCharmURL.Schema) {
+			return nil, errors.Errorf("cannot specify revision in a charm or bundle name. Please use --revision.")
+		}
+		if d.revision != -1 && charm.CharmHub.Matches(userCharmURL.Schema) && d.channel.Empty() {
+			// Tell the user they need to specify a channel
+			return nil, errors.Errorf("specifying a revision requires a channel for future upgrades. Please use --channel")
+		}
+
+		if d.revision != -1 {
+			userCharmURL = userCharmURL.WithRevision(d.revision)
+		}
+
+		platform := utils.MakePlatform(d.constraints, d.base, d.modelConstraints)
+		origin, err := utils.DeduceOrigin(userCharmURL, d.channel, platform)
+		if err != nil {
 			return nil, errors.Trace(err)
-		} else if deploy != nil {
-			return deploy, nil
+		}
+
+		// TODO (cderici): check the validity of the comment below
+		// Resolve the bundle URL using the channel supplied via the channel
+		// supplied. All charms within this bundle unless pinned via a channel are
+		// NOT expected to be in the same channel as the bundle channel.
+		// The pinning of a bundle does not flow down to charms as well. Each charm
+		// has it's own channel supplied via a bundle, if no is supplied then the
+		// channel is worked out via the resolving what is available.
+		// See: LP:1677404 and LP:1832873
+		bundleURL, bundleOrigin, bundleResolveErr := resolver.ResolveBundleURL(userCharmURL, origin)
```

Comment:
ResolveBundleURL is a wrapper on ResolveCharm which checks the type in origin. If you use ResolveCharm here, check the type, you can keep the origin already resolved upstream for both the charm and the bundle.